internal/upload/uploader/uploader.go

```diff
@@ -98,16 +103,22 @@ type (
 	}
 )
 
-func New(ctx context.Context, cl locking.ContractLocker, cs ContractStore, hm host.HostManager, c api.ContractMetadata, l *zap.SugaredLogger) *Uploader {
+func New(ctx context.Context, cl locking.ContractLocker, cs ContractStore, hm host.HostManager, hs HostStore, c api.ContractMetadata, l *zap.SugaredLogger) (*Uploader, error) {
+	host, err := hs.Host(ctx, c.HostKey)
+	if err != nil {
+		return nil, err
+	}
+	siamuxAddr := host.Info().SiamuxAddr
 	return &Uploader{
 		cl:     cl,
 		cs:     cs,
 		hm:     hm,
+		hs:     hs,
 		logger: l,
 
 		// static
 		hk:              c.HostKey,
-		siamuxAddr:      c.SiamuxAddr,
+		siamuxAddr:      siamuxAddr,
```

Comment:
Can you check if `SiamuxAddr` is even used? If not, you can remove the field and avoid fetching the host here. You also don't need the `HostStore` then.

Reply:
Do we need the siamux addr for this part?

```
	return &Uploader{
...
        host:      hm.Host(c.HostKey, c.ID, c.SiamuxAddr),
...
}
```

refreshUploaders is called by newUpload which is called by UploadPackedSlab and UploadShards.  Maybe I'm missing something but I couldn't find any host info in that call stack, so it seems like we'll have to get info from the host store at some point.

Reply:
True, we do have to get that info at some point, but it's better to get all the host info at once as opposed to fetching on the fly every time. I would probably add 
```go
type hostContract struct {
    api.ContractMetadata
    api.HostInfo
}
```
which we manually construct as high up as possible. Where we fetch contracts we can fetch usable hosts and zip those two.

Reply:
@peterjan I'd propose that we keep this as-is for now. Sure we need to fetch the host now too every time we previously only fetched the contract but imo it's not worth the refactor to worry about that now.
After the hardfork we will be able to tear all of that contract renewal safety out anyway.

Reply:
Hm ok. I knew more or less how this could be implemented because I dove into it a bit for RHP4 uploads. Matter of fact the RHP4 uploads PR _does_ introduce this combined type and in the process I also had to get rid of the `HostIP` and `SiamuxAddr`.